pkg/crmutil/controller-data.go

```diff
@@ -336,12 +370,41 @@ func (cd *CRMHandler) CloudletHasTrustPolicy(ctx context.Context, cloudletKey *e
 	return true, nil
 }
 
+func (cd *CRMHandler) appInstDNSChanged(ctx context.Context, pf platform.Platform, app *edgeproto.App, oldURI string, new *edgeproto.AppInst, sender edgeproto.AppInstInfoSender) (reterr error) {
+	var err error
+
+	updateAppCacheCallback := sender.SendStatusIgnoreErr
+	log.SpanLog(ctx, log.DebugLevelInfra, "AppInstDNSChanged", "key", new.Key, "old dns", oldURI, "new dns", new.Uri)
+
+	// reset status messages
+	err = sender.SendState(edgeproto.TrackedState_UPDATING, edgeproto.WithSenderResetStatus())
+	if err != nil {
+		return err
+	}
+	err = pf.ChangeAppInstDNS(ctx, app, new, oldURI, updateAppCacheCallback)
+	if err != nil {
+		err := fmt.Errorf("update failed: %s", err)
+		sender.SendState(edgeproto.TrackedState_UPDATE_ERROR, edgeproto.WithStateError(err))
+		return err
+	}
+	log.SpanLog(ctx, log.DebugLevelInfra, "AppInstDNSChanged done", "key", new.Key, "old dns", oldURI, "new dns", new.Uri)
+	sender.SendState(edgeproto.TrackedState_READY)
+	return nil
+}
+
 func (cd *CRMHandler) AppInstChanged(ctx context.Context, target *edgeproto.CloudletKey, new *edgeproto.AppInst, sender edgeproto.AppInstInfoSender) (nu NeedsUpdate, reterr error) {
+	var err error
+
 	log.SpanLog(ctx, log.DebugLevelInfra, "appInstChanged", "new", new)
 
-	var err error
 	fmap := edgeproto.MakeFieldMap(new.Fields)
-	if !fmap.Has(edgeproto.AppInstFieldState) {
+	dnsUpdate := false
+	oldURI, ok := new.Annotations[cloudcommon.AnnotationPreviousDNSName]
+	if fmap.Has(edgeproto.AppInstFieldUri) && ok && oldURI != new.Uri {
+		dnsUpdate = true
+	}
+
```

Comment:
Same thing here, just check for the field flag, since the annotation is always different from the new.Uri.